subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java

```diff
@@ -475,26 +472,50 @@ public <T extends Task> T createWithoutConstructor(String name, Class<T> type, l
     }
 
     @Override
-    public Task findByPath(String path) {
-        Path.validatePath(path);
-        if (!path.contains(Project.PATH_SEPARATOR)) {
-            return findByName(path);
+    public @Nullable Task findByPath(String pathStr) {
+        Path path = Path.path(pathStr);
+        if (!path.isAbsolute() && path.getParent() == null) {
+            return findByName(pathStr);
         }
 
-        String projectPath = StringUtils.substringBeforeLast(path, Project.PATH_SEPARATOR);
-        String projectPathOrRoot = Strings.isNullOrEmpty(projectPath) ? Project.PATH_SEPARATOR : projectPath;
-        ProjectInternal project = projectRegistry.getProject(this.project.absoluteProjectPath(projectPathOrRoot));
-        if (project == null) {
-            return null;
-        }
-        project.getOwner().ensureTasksDiscovered();
+        Path projectPath = path.getParent();
+        Path targetProjectPath = projectPath != null
+            ? this.project.getProjectIdentity().getProjectPath().absolutePath(projectPath)
+            : Path.ROOT;
+        String targetTaskName = path.getName();
 
-        return project.getTasks().findByName(StringUtils.substringAfterLast(path, Project.PATH_SEPARATOR));
+        return findTaskInProject(targetProjectPath, targetTaskName);
     }
 
-    @Override
-    public Task resolveTask(String path) {
-        return getByPath(path);
+    /**
+     * Find a task with the given name in the project with the given path.
+     * <p>
+     * TODO: A task container in one project has no business providing the tasks of
```

Comment:
💭 Do we have an issue we can link to?

Reply:
Created https://github.com/gradle/gradle/issues/34939